

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Pv4 heterozygous calls &#8212; Genomic transmission graph</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Pv4_het_calls_230318';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="_static/logo.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="index.html">
                    Welcome
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="how-to-use-this-book.html">How to use this book</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="using-google-colab.html">Using Google Colab</a></li>
<li class="toctree-l2"><a class="reference internal" href="glossary.html">Glossary</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="underlying-concepts.html">Underlying concepts</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="basic-model.html">The basic model</a></li>
<li class="toctree-l2"><a class="reference internal" href="coalescence-time-basic.html">Time to coalescence</a></li>
<li class="toctree-l2"><a class="reference internal" href="nucleotide-diversity.html">Nucleotide diversity</a></li>
<li class="toctree-l2"><a class="reference internal" href="haplotype-homozygosity.html">Haplotype homozygosity</a></li>
<li class="toctree-l2"><a class="reference internal" href="ibd.html">Identity by descent</a></li>
<li class="toctree-l2"><a class="reference internal" href="migration-simple.html">Effects of migration</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="within-host.html">Within-host diversity</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="within-host-pi.html">Within-host <span class="math notranslate nohighlight">\(\pi_W\)</span></a></li>
<li class="toctree-l2"><a class="reference internal" href="hw-hs.html"><em>Hw - Hs</em> relationship</a></li>
<li class="toctree-l2"><a class="reference internal" href="fws-compare-methods.html">Rate of superinfection</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="coalestr-features.html">How to use <code class="docutils literal notranslate"><span class="pre">coalestr</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="create-population.html">Create a <code class="docutils literal notranslate"><span class="pre">Population</span></code></a></li>



<li class="toctree-l2"><a class="reference internal" href="get-coalescent.html"><code class="docutils literal notranslate"><span class="pre">get_coalescent</span></code></a></li>


<li class="toctree-l2"><a class="reference internal" href="get-diversity.html"><code class="docutils literal notranslate"><span class="pre">get_diversity</span></code></a></li>

<li class="toctree-l2"><a class="reference internal" href="create-subpopulation.html">Create a subpopulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="time-series.html">Analysing a time series</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="epidemiological-scenarios.html">Epidemiological scenarios</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="species_history.html">Create a metapopulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="scenario_1.html">Migration time series</a></li>
<li class="toctree-l2"><a class="reference internal" href="local_outbreak.html">Modelling a local outbreak</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="draft-ms.html">Draft ms files</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-launch-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Launch interactive content">
    <i class="fas fa-rocket"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://colab.research.google.com/github/d-kwiat/gtg/blob/main/Pv4_het_calls_230318.ipynb" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch onColab"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img src="_static/images/logo_colab.png">
  </span>
<span class="btn__text-container">Colab</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/d-kwiat/gtg" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/d-kwiat/gtg/issues/new?title=Issue%20on%20page%20%2FPv4_het_calls_230318.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/Pv4_het_calls_230318.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Pv4 heterozygous calls</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#extract-lists-of-samples-and-variant-properties-as-arrays">Extract lists of samples and variant properties as arrays</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#select-high-quality-biallelic-snps">Select high-quality biallelic SNPs</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#save-allele-depths-for-selected-snps-in-allele-depth-1-npy">Save allele depths for selected SNPs in <code class="docutils literal notranslate"><span class="pre">allele_depth_1.npy</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#select-samples-and-list-region-of-each-selected-sample">Select samples and list region of each selected sample</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#copy-allele-depth-for-qc-pass-samples-from-allele-depth-1-into-allele-depth-2">Copy allele depth for QC pass samples from <code class="docutils literal notranslate"><span class="pre">allele_depth_1</span></code> into <code class="docutils literal notranslate"><span class="pre">allele_depth_2</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#calculate-within-host-heterozygosity-for-each-snp-and-save-as-wiho-het-npy">Calculate within-host heterozygosity for each SNP and save as <code class="docutils literal notranslate"><span class="pre">wiho_het.npy</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#calculate-mean-within-host-heterozygosity-for-each-snp">Calculate mean within-host heterozygosity for each SNP</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#create-snp-filter-to-filter-out-possible-hyperhets">Create <code class="docutils literal notranslate"><span class="pre">snp_filter</span></code> to filter out possible hyperhets</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#create-sample-filter-to-select-a-particular-geographical-region">Create <code class="docutils literal notranslate"><span class="pre">sample_filter</span></code> to select a particular geographical region</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#calculate-mean-heterozygosity-per-sample-after-filtering-samples-and-snps">Calculate mean heterozygosity per sample after filtering samples and SNPs</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#plot-histogram-of-within-host-nucleotide-diversity">Plot histogram of within-host nucleotide diversity</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section id="pv4-heterozygous-calls">
<h1>Pv4 heterozygous calls<a class="headerlink" href="#pv4-heterozygous-calls" title="Permalink to this heading">#</a></h1>
<p>My Drive/GitHub/gtg/Pf4_het_calls_230318.ipynb</p>
<p>This code examines the frequency distribution of heterozgous genotype calls in the <a class="reference external" href="https://www.malariagen.net/sites/default/files/File8_Pf_6_README_20191010_0.txt">Pv4 dataset</a>.  It uses downloaded Zarr data and sample metadata that are stored on Google Drive\My Drive\Colab Data\Pv4.</p>
<ul class="simple">
<li><p>extract lists of samples (n = 1,895) and variant properties (n = 4,571,056) as arrays</p></li>
<li><p>select variants that are high-quality biallelic coding SNPs (n = 429,612)</p></li>
<li><p>extract allele depth for high-quality SNPs from Zarr into <code class="docutils literal notranslate"><span class="pre">allele_depth_1.npy</span></code>.  We do this in chunks of 10,000 variants at a time to avoid memory problems</p></li>
<li><p>extract sample metadata from samples.txt into a dataframe</p></li>
<li><p>read info about sample QC into an array</p></li>
<li><p>extract allele depth for high-quality SNPs and high-quality samples (n = 1,072) into <code class="docutils literal notranslate"><span class="pre">allele_depth_2.npy</span></code></p></li>
<li><p>calculate within-host heterozygosity for each SNP in each sample and save to <code class="docutils literal notranslate"><span class="pre">wiho_het.npy</span></code></p></li>
<li><p>calculate mean within-host heterozygosity for each SNP across all samples and save to a list called wiho_het_per_snp</p></li>
<li><p>using wiho_het_per_snp, create <code class="docutils literal notranslate"><span class="pre">snp_filter</span></code> to filter out potential hyperhet SNPs</p></li>
<li><p>create <code class="docutils literal notranslate"><span class="pre">sample_filter</span></code> to select samples from a particular geographical region</p></li>
<li><p>calculate <span class="math notranslate nohighlight">\(\pi_W\)</span> for each sample, using mean within-host heterozygosity per sample, and allowing for the number of SNPs ascertained and the size of coding regions</p></li>
<li><p>plot a histogram of nucleotide diversity per sample</p></li>
<li><p>view the histogram for different filter cutoffs for hyperhet SNPs and for different regions e.g. West Africa vs Southeast Asia</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">google.colab</span> <span class="kn">import</span> <span class="n">drive</span>

<span class="n">drive</span><span class="o">.</span><span class="n">mount</span><span class="p">(</span><span class="s1">&#39;/content/drive&#39;</span><span class="p">)</span>

<span class="n">data_path</span> <span class="o">=</span> <span class="s2">&quot;/content/drive/My Drive/Colab Data/Pv4/&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>pip<span class="w"> </span>install<span class="w"> </span>zarr

<span class="kn">import</span> <span class="nn">zarr</span>
<span class="kn">import</span> <span class="nn">numcodecs</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">statistics</span> <span class="k">as</span> <span class="nn">stat</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_path</span> <span class="o">=</span> <span class="s1">&#39;G:</span><span class="se">\\</span><span class="s1">My Drive</span><span class="se">\\</span><span class="s1">Colab Data</span><span class="se">\\</span><span class="s1">Pv4</span><span class="se">\\</span><span class="s1">&#39;</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">statistics</span> <span class="k">as</span> <span class="nn">stat</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">callset</span> <span class="o">=</span> <span class="n">zarr</span><span class="o">.</span><span class="n">open_group</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s1">&#39;Pv_4.zarr&#39;</span><span class="p">,</span> <span class="n">mode</span> <span class="o">=</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
<span class="c1"># callset.tree()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "9f52de55022f4a108612dcfc7df6d9ad", "version_major": 2, "version_minor": 0}</script></div>
</div>
<section id="extract-lists-of-samples-and-variant-properties-as-arrays">
<h2>Extract lists of samples and variant properties as arrays<a class="headerlink" href="#extract-lists-of-samples-and-variant-properties-as-arrays" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>1,895 samples</p></li>
<li><p>4,571,056 variants</p></li>
</ul>
</section>
<section id="select-high-quality-biallelic-snps">
<h2>Select high-quality biallelic SNPs<a class="headerlink" href="#select-high-quality-biallelic-snps" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>high quality means ‘filter_pass == True’ &amp; ‘vqslod &gt; 3’</p></li>
<li><p>429,612 variants meet criteria</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">select</span></code> is a mask for an array of 4,571,056 variants</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">samples</span> <span class="o">=</span> <span class="n">callset</span><span class="p">[</span><span class="s1">&#39;samples&#39;</span><span class="p">][:]</span>

<span class="n">filter_pass</span> <span class="o">=</span> <span class="n">callset</span><span class="p">[</span><span class="s1">&#39;variants/FILTER_PASS&#39;</span><span class="p">][:]</span>
<span class="n">vqslod</span> <span class="o">=</span> <span class="n">callset</span><span class="p">[</span><span class="s1">&#39;variants/VQSLOD&#39;</span><span class="p">][:]</span>
<span class="n">snp</span> <span class="o">=</span> <span class="n">callset</span><span class="p">[</span><span class="s1">&#39;variants/is_snp&#39;</span><span class="p">][:]</span>
<span class="n">coding</span> <span class="o">=</span> <span class="n">callset</span><span class="p">[</span><span class="s1">&#39;variants/CDS&#39;</span><span class="p">][:]</span>
<span class="n">numalt</span> <span class="o">=</span> <span class="n">callset</span><span class="p">[</span><span class="s1">&#39;variants/numalt&#39;</span><span class="p">][:]</span>
<span class="n">effect</span> <span class="o">=</span> <span class="n">callset</span><span class="p">[</span><span class="s1">&#39;variants/SNPEFF_EFFECT&#39;</span><span class="p">][:]</span>

<span class="c1"># select biallelic synonymous coding SNPs and create a mask</span>
<span class="n">select</span> <span class="o">=</span> <span class="p">(</span><span class="n">filter_pass</span> <span class="o">==</span> <span class="kc">True</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">snp</span> <span class="o">==</span> <span class="kc">True</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">numalt</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">vqslod</span> <span class="o">&gt;</span> <span class="mi">3</span><span class="p">)</span>

<span class="c1"># &amp; (coding == True) &amp; (effect == &#39;SYNONYMOUS_CODING&#39;) </span>

<span class="n">selected_snps</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">select</span> <span class="o">==</span> <span class="kc">True</span><span class="p">)</span>

<span class="n">total_variants</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">snp</span><span class="p">)</span>
<span class="n">total_samples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;There are </span><span class="si">{0:d}</span><span class="s1"> variants and </span><span class="si">{1:d}</span><span class="s1"> selected SNPs in </span><span class="si">{2:d}</span><span class="s1"> samples.&#39;</span><span class="o">.</span>
      <span class="nb">format</span><span class="p">(</span><span class="n">total_variants</span><span class="p">,</span> <span class="n">selected_snps</span><span class="p">,</span> <span class="n">total_samples</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>There are 6051696 variants and 502221 selected SNPs in 7113 samples.
</pre></div>
</div>
</div>
</div>
</section>
<section id="save-allele-depths-for-selected-snps-in-allele-depth-1-npy">
<h2>Save allele depths for selected SNPs in <code class="docutils literal notranslate"><span class="pre">allele_depth_1.npy</span></code><a class="headerlink" href="#save-allele-depths-for-selected-snps-in-allele-depth-1-npy" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>Use the <code class="docutils literal notranslate"><span class="pre">select</span></code> mask created above to pull out selected SNPs</p></li>
<li><p>Copy allele depths for selected SNPs from ‘calldata/AD’ into <code class="docutils literal notranslate"><span class="pre">allele_depth_1</span></code></p>
<ul>
<li><p>calldata/AD allows for multiallelic loci by containing allele depth for up to 7 variant alleles</p></li>
<li><p>Since we focus on biallelic SNPs we include only the first two alleles from calldata/AD</p></li>
</ul>
</li>
<li><p>This file is 3.2GB</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># We extract allele depth data for selected SNPs into a numpy array.</span>
<span class="c1"># We do this in chunks of 10000 variants at a time to avoid memory problems.</span>
<span class="c1"># This step can be omitted by retrieving saved data in the next section.</span>

<span class="n">ad</span> <span class="o">=</span> <span class="n">callset</span><span class="p">[</span><span class="s1">&#39;calldata/AD&#39;</span><span class="p">]</span>

<span class="n">allele_depth_1</span> <span class="o">=</span> <span class="mi">9</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="n">selected_snps</span><span class="p">,</span> <span class="n">total_samples</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int16</span><span class="p">)</span>

<span class="n">selected_snp_id</span> <span class="o">=</span> <span class="mi">0</span>

<span class="n">chunk_size</span> <span class="o">=</span> <span class="mi">10000</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">total_variants</span> <span class="o">/</span> <span class="n">chunk_size</span><span class="p">)):</span>
    
    <span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    
    <span class="n">chunk</span> <span class="o">=</span> <span class="n">ad</span><span class="p">[(</span><span class="n">i</span> <span class="o">*</span> <span class="n">chunk_size</span><span class="p">)</span> <span class="p">:</span> <span class="p">((</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">chunk_size</span><span class="p">),</span> <span class="p">:,</span> <span class="p">:</span><span class="mi">2</span><span class="p">]</span>
    
    <span class="n">chunk_snps</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">chunk_size</span><span class="p">):</span>
        
        <span class="n">variant_id</span> <span class="o">=</span> <span class="n">i</span> <span class="o">*</span> <span class="n">chunk_size</span> <span class="o">+</span> <span class="n">j</span>
    
        <span class="k">if</span> <span class="n">select</span><span class="p">[</span><span class="n">variant_id</span><span class="p">]</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
            
            <span class="n">allele_depth_1</span><span class="p">[</span><span class="n">selected_snp_id</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">chunk</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span>
            
            <span class="n">selected_snp_id</span> <span class="o">+=</span> <span class="mi">1</span>
            
            <span class="n">chunk_snps</span> <span class="o">+=</span> <span class="mi">1</span>
               
    <span class="n">chunk_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start_time</span>
        
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;chunk </span><span class="si">{0:d}</span><span class="s1"> had </span><span class="si">{1:d}</span><span class="s1"> SNPs and took </span><span class="si">{2:.2f}</span><span class="s1"> seconds&#39;</span><span class="o">.</span>
          <span class="nb">format</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">chunk_snps</span><span class="p">,</span> <span class="n">chunk_time</span><span class="p">))</span>

<span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

<span class="n">chunk</span> <span class="o">=</span> <span class="n">ad</span><span class="p">[(</span><span class="n">i</span> <span class="o">*</span> <span class="n">chunk_size</span><span class="p">)</span> <span class="p">:,</span> <span class="p">:,</span> <span class="p">:]</span>

<span class="n">chunk_snps</span> <span class="o">=</span> <span class="mi">0</span>

<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">total_variants</span> <span class="o">%</span> <span class="n">chunk_size</span><span class="p">):</span>
    
    <span class="k">if</span> <span class="n">select</span><span class="p">[(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">chunk_size</span> <span class="o">+</span> <span class="n">j</span><span class="p">]</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
            
        <span class="n">allele_depth_1</span><span class="p">[</span><span class="n">selected_snp_id</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">chunk</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span>
        
        <span class="n">selected_snp_id</span> <span class="o">+=</span> <span class="mi">1</span>
            
        <span class="n">chunk_snps</span> <span class="o">+=</span> <span class="mi">1</span>
               
    <span class="n">chunk_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start_time</span>
        
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;chunk </span><span class="si">{0:d}</span><span class="s1"> had </span><span class="si">{1:d}</span><span class="s1"> SNPs and took </span><span class="si">{2:.1f}</span><span class="s1"> seconds&#39;</span><span class="o">.</span>
     <span class="nb">format</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">chunk_snps</span><span class="p">,</span> <span class="n">chunk_time</span><span class="p">))</span>

<span class="n">total_snps</span> <span class="o">=</span> <span class="n">selected_snp_id</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;total of </span><span class="si">{0:d}</span><span class="s1"> selected snps&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">total_snps</span><span class="p">))</span>

<span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s1">&#39;220318_allele_depth_1&#39;</span><span class="p">,</span> <span class="n">allele_depth_1</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="select-samples-and-list-region-of-each-selected-sample">
<h2>Select samples and list region of each selected sample<a class="headerlink" href="#select-samples-and-list-region-of-each-selected-sample" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>‘Population’ in Pv4_samples.txt gives: LAM=Latin America, AF=Africa, WAS=Western Asia, WSEA=Western SE Asia, ESEA=Eastern SE Asia, MSEA=Maritime SE Asia, OCE=Oceania, unassigned</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s1">&#39;samples.txt&#39;</span><span class="p">)</span>

<span class="n">select_sample</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;QC pass&#39;</span><span class="p">])</span>

<span class="n">selected_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">select_sample</span> <span class="o">==</span> <span class="kc">True</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;There are&#39;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">selected_samples</span><span class="p">),</span> <span class="s1">&#39;selected samples that passed QC&#39;</span><span class="p">)</span>

<span class="n">region</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">total_samples</span><span class="p">):</span>
    
    <span class="k">if</span> <span class="n">select_sample</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        
        <span class="n">region</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Population&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">])</span>
        
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s2">&quot;230318_region&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fp</span><span class="p">:</span>
    <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">region</span><span class="p">,</span> <span class="n">fp</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>There are 5970 selected samples that passed QC
</pre></div>
</div>
</div>
</div>
</section>
<section id="copy-allele-depth-for-qc-pass-samples-from-allele-depth-1-into-allele-depth-2">
<h2>Copy allele depth for QC pass samples from <code class="docutils literal notranslate"><span class="pre">allele_depth_1</span></code> into <code class="docutils literal notranslate"><span class="pre">allele_depth_2</span></code><a class="headerlink" href="#copy-allele-depth-for-qc-pass-samples-from-allele-depth-1-into-allele-depth-2" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>‘QC pass’ field in samples.txt gives True/False</p></li>
<li><p>Save as .npy file (~11.7GB) so that this step only has to be run once</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">allele_depth_1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s1">&#39;230318_allele_depth_1.npy&#39;</span><span class="p">)</span>

<span class="n">allele_depth_2</span> <span class="o">=</span> <span class="mi">9</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="n">selected_snps</span><span class="p">,</span> <span class="n">selected_samples</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int16</span><span class="p">)</span>

<span class="n">selected_sample_id</span> <span class="o">=</span> <span class="mi">0</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">total_samples</span><span class="p">):</span>
    
    <span class="k">if</span> <span class="n">select_sample</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        
        <span class="n">allele_depth_2</span><span class="p">[:,</span> <span class="n">selected_sample_id</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">allele_depth_1</span><span class="p">[:,</span> <span class="n">i</span><span class="p">,</span> <span class="p">:]</span>
        
        <span class="n">selected_sample_id</span> <span class="o">+=</span> <span class="mi">1</span>
        
<span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s1">&#39;230318_allele_depth_2&#39;</span><span class="p">,</span> <span class="n">allele_depth_2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>There are 5970 selected samples that passed QC
</pre></div>
</div>
</div>
</div>
</section>
<section id="calculate-within-host-heterozygosity-for-each-snp-and-save-as-wiho-het-npy">
<h2>Calculate within-host heterozygosity for each SNP and save as <code class="docutils literal notranslate"><span class="pre">wiho_het.npy</span></code><a class="headerlink" href="#calculate-within-host-heterozygosity-for-each-snp-and-save-as-wiho-het-npy" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>if ref_allele + alt_allele == 1 then write het = 0.</p></li>
<li><p>this file is 3.6GB because dtype=float</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">allele_depth_2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s1">&#39;230318_allele_depth_2.npy&#39;</span><span class="p">)</span> <span class="c1">#, mmap_mode = &#39;r&#39;)</span>

<span class="n">selected_samples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">allele_depth_2</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:,</span> <span class="mi">0</span><span class="p">])</span>

<span class="n">wiho_het</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="n">selected_snps</span><span class="p">,</span> <span class="n">selected_samples</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">selected_snps</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">selected_samples</span><span class="p">):</span>
        
        <span class="n">ref_allele</span> <span class="o">=</span> <span class="n">allele_depth_2</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
        <span class="n">alt_allele</span> <span class="o">=</span> <span class="n">allele_depth_2</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
        <span class="n">denominator</span> <span class="o">=</span> <span class="n">ref_allele</span> <span class="o">+</span> <span class="n">alt_allele</span>
        <span class="k">if</span> <span class="n">denominator</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">allele_freq</span> <span class="o">=</span> <span class="n">alt_allele</span> <span class="o">/</span> <span class="n">denominator</span>
            <span class="n">het</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">allele_freq</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">allele_freq</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">het</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># this is a hack</span>
            
        <span class="n">wiho_het</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">het</span>

    <span class="k">if</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">100</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        
<span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s1">&#39;230318_wiho_het.npy&#39;</span><span class="p">,</span> <span class="n">wiho_het</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="calculate-mean-within-host-heterozygosity-for-each-snp">
<h2>Calculate mean within-host heterozygosity for each SNP<a class="headerlink" href="#calculate-mean-within-host-heterozygosity-for-each-snp" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">wiho_het</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s1">&#39;230318_wiho_het.npy&#39;</span><span class="p">,</span> <span class="n">mmap_mode</span> <span class="o">=</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>

<span class="n">het_per_snp</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">selected_snps</span><span class="p">):</span>
    
    <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">wiho_het</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:])</span>
    
    <span class="n">het_per_snp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s2">&quot;230318_het_per_snp&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">filepath</span><span class="p">:</span>
    <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">het_per_snp</span><span class="p">,</span> <span class="n">filepath</span><span class="p">)</span>        
</pre></div>
</div>
</div>
</div>
</section>
<section id="create-snp-filter-to-filter-out-possible-hyperhets">
<h2>Create <code class="docutils literal notranslate"><span class="pre">snp_filter</span></code> to filter out possible hyperhets<a class="headerlink" href="#create-snp-filter-to-filter-out-possible-hyperhets" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>Here we use het &gt;= 0.02 as our criterion for exclusion</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s2">&quot;230318_het_per_snp&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fp</span><span class="p">:</span>
    <span class="n">het_per_snp</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">fp</span><span class="p">)</span>

<span class="n">snp_filter</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">selected_snps</span><span class="p">):</span>
    
    <span class="k">if</span> <span class="n">het_per_snp</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.02</span><span class="p">:</span>
        
        <span class="n">snp_filter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="k">else</span><span class="p">:</span>
        <span class="n">snp_filter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        
<span class="n">filtered_snps</span> <span class="o">=</span> <span class="n">snp_filter</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;This includes&#39;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">filtered_snps</span><span class="p">),</span> <span class="s1">&#39;SNPs&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>This includes 496459 SNPs
</pre></div>
</div>
</div>
</div>
</section>
<section id="create-sample-filter-to-select-a-particular-geographical-region">
<h2>Create <code class="docutils literal notranslate"><span class="pre">sample_filter</span></code> to select a particular geographical region<a class="headerlink" href="#create-sample-filter-to-select-a-particular-geographical-region" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>LAM=Latin America, AF=Africa, WAS=Western Asia, WSEA=Western SE Asia, ESEA=Eastern SE Asia, MSEA=Maritime SE Asia, OCE=Oceania, unassigned</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s2">&quot;230318_region&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fp</span><span class="p">:</span>
    <span class="n">region</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">fp</span><span class="p">)</span>

<span class="n">sample_filter</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">selected_samples</span><span class="p">):</span>
    
    <span class="k">if</span> <span class="n">region</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="kc">False</span><span class="p">:</span>
        
        <span class="n">sample_filter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        
    <span class="k">else</span><span class="p">:</span>
        
        <span class="n">sample_filter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        
<span class="n">filtered_samples</span> <span class="o">=</span> <span class="n">sample_filter</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;This includes&#39;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">filtered_samples</span><span class="p">),</span> <span class="s1">&#39;samples&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>This includes 5970 samples
</pre></div>
</div>
</div>
</div>
</section>
<section id="calculate-mean-heterozygosity-per-sample-after-filtering-samples-and-snps">
<h2>Calculate mean heterozygosity per sample after filtering samples and SNPs<a class="headerlink" href="#calculate-mean-heterozygosity-per-sample-after-filtering-samples-and-snps" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>Creates a list of het values, one for each selected &amp; filtered sample</p></li>
<li><p>This is the mean het for selected &amp; filtered SNPs and not for all nucleotide positions</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">wiho_het</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s1">&#39;230318_wiho_het.npy&#39;</span><span class="p">,</span> <span class="n">mmap_mode</span> <span class="o">=</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>

<span class="n">het_per_sample</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">selected_samples</span><span class="p">):</span>
    
    <span class="k">if</span> <span class="n">sample_filter</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>

        <span class="n">het</span> <span class="o">=</span> <span class="p">[]</span>
    
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">selected_snps</span><span class="p">):</span>
    
            <span class="k">if</span> <span class="n">snp_filter</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span> 
            
                <span class="n">het</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">wiho_het</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="n">i</span><span class="p">])</span>
    
        <span class="n">y</span> <span class="o">=</span> <span class="n">stat</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">het</span><span class="p">)</span>
    
        <span class="n">het_per_sample</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">1000</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s2">&quot;230318_het_per_sample_ALL&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">filepath</span><span class="p">:</span>
    <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">het_per_sample</span><span class="p">,</span> <span class="n">filepath</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0
1000
2000
3000
4000
5000
</pre></div>
</div>
</div>
</div>
</section>
<section id="plot-histogram-of-within-host-nucleotide-diversity">
<h2>Plot histogram of within-host nucleotide diversity<a class="headerlink" href="#plot-histogram-of-within-host-nucleotide-diversity" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># All regions - only SNPs with maf &lt; 0.01</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s2">&quot;230318_het_per_sample_ALL&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">filepath</span><span class="p">:</span>
    <span class="n">het_per_sample</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>
    
<span class="n">selected_snps</span> <span class="o">=</span> <span class="mi">494829</span> <span class="c1"># this is &lt;0.01 maf, total is 502221</span>
<span class="n">coding_positions</span> <span class="o">=</span> <span class="mi">12028350</span> <span class="c1"># from Gardner et al 2002</span>

<span class="n">convert_het_to_pi</span> <span class="o">=</span> <span class="n">selected_snps</span> <span class="o">/</span> <span class="n">coding_positions</span>

<span class="n">het_per_sample_nonzero</span> <span class="o">=</span> <span class="p">[</span><span class="n">convert_het_to_pi</span> <span class="o">*</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">het_per_sample</span> <span class="k">if</span> <span class="n">x</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">]</span>
<span class="n">log_het_per_sample</span> <span class="o">=</span> <span class="p">[</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">het_per_sample_nonzero</span><span class="p">]</span>
<span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">log_het_per_sample</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="mi">100</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Within-host nucleotide diversity </span><span class="se">\u03C0</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Number of samples&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/fd1b07f1fc5577afa2bcb06d48e769f8b91a41bc16c017bce2a09dc67a22d9be.png" src="_images/fd1b07f1fc5577afa2bcb06d48e769f8b91a41bc16c017bce2a09dc67a22d9be.png" />
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#extract-lists-of-samples-and-variant-properties-as-arrays">Extract lists of samples and variant properties as arrays</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#select-high-quality-biallelic-snps">Select high-quality biallelic SNPs</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#save-allele-depths-for-selected-snps-in-allele-depth-1-npy">Save allele depths for selected SNPs in <code class="docutils literal notranslate"><span class="pre">allele_depth_1.npy</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#select-samples-and-list-region-of-each-selected-sample">Select samples and list region of each selected sample</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#copy-allele-depth-for-qc-pass-samples-from-allele-depth-1-into-allele-depth-2">Copy allele depth for QC pass samples from <code class="docutils literal notranslate"><span class="pre">allele_depth_1</span></code> into <code class="docutils literal notranslate"><span class="pre">allele_depth_2</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#calculate-within-host-heterozygosity-for-each-snp-and-save-as-wiho-het-npy">Calculate within-host heterozygosity for each SNP and save as <code class="docutils literal notranslate"><span class="pre">wiho_het.npy</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#calculate-mean-within-host-heterozygosity-for-each-snp">Calculate mean within-host heterozygosity for each SNP</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#create-snp-filter-to-filter-out-possible-hyperhets">Create <code class="docutils literal notranslate"><span class="pre">snp_filter</span></code> to filter out possible hyperhets</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#create-sample-filter-to-select-a-particular-geographical-region">Create <code class="docutils literal notranslate"><span class="pre">sample_filter</span></code> to select a particular geographical region</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#calculate-mean-heterozygosity-per-sample-after-filtering-samples-and-snps">Calculate mean heterozygosity per sample after filtering samples and SNPs</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#plot-histogram-of-within-host-nucleotide-diversity">Plot histogram of within-host nucleotide diversity</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Dominic Kwiatkowski
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>