{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Within-host diversity\n",
    "\n",
    "Within-host diversity, i.e. the genetic diversity of the parasite population within an infected individual, is determined by a number of different factors.  *Superinfection* gives rise to multiclonal infections and is an important driver of within-host diversity. In our model the rate of superinfection is given by $\\chi$.\n",
    "\n",
    "Following an episode of superinfection, a multiclonal infection may propagate along a transmission chain for several generations - this is known as *cotransmission*.  Sexual recombination occurs with each generation of cotransmission, so that the parasites become increasing related to each other although they remain genetically diverse.  Cotransmission is more likely to continue over multiple generations if the quantum of transmission $Q$ is large, and it will not occur if $Q=1$.  \n",
    "\n",
    "*Mutation* and *recombination* also contribute to within-host diversity but at a much lower level than superinfection and cotransmission.  Again this depends on $Q$ which determines the force of genetic drift in eliminating the diversity generated by mutation, recombination and superinfection.\n",
    "\n",
    "Within-host diversity can be empirically quantified by a number of approaches:\n",
    "\n",
    "* Complexity of infection (COI).  This involves estimating the number of distinct parasite genotypes observed in an infected individual using genetic barcodes, microsatellites, microhaplotyes and other sequence data.\n",
    "\n",
    "* [Measurement of within-host nucleotide diversity ($\\pi_W$)](within-host-pi.ipynb) by deep genome sequencing of individual infections.\n",
    "\n",
    "* [Estimation of the inbreeding coefficient $F_{WS}$](fws-compare-methods.ipynb) by comparing mean within-host heterozygosity $\\widehat{H}_W$ with local subpopulation heterozygosity $H_S$ across hundreds of thousands of SNPs.\n",
    "\n",
    "* [Measurement of the proportion of the genome that is identical by descent (IBD)](ibd.ipynb) between alleles sampled within a host. In our model this is approximated by $\\gamma_W$, the within-host haplotype homozygosity of a 2 centimorgan locus.  This is difficult to measure using conventional short-read sequencing but is becoming possible due to recent advances in single-cell sequencing, long-read sequencing and algorithmic construction of haplotypes from short-read data.\n",
    "\n",
    "The genomic transmission graph provides a number of fundamental insights into the driving forces of within-host diversity, and also into how measurements of $\\pi_W$ and $F_{WS}$ can be used to estimate the transmission parameters $Q$ and $\\chi$.\n",
    "\n",
    "The worked example below shows how to obtain $\\pi_W$ (mean within-host nucleotide diversity) and $\\gamma_W$ (within-host haplotype homozygosity of a 2 centimorgan locus) for different combinations of transmission parameters using `coalestr`.\n",
    "\n",
    "On other pages we examine:\n",
    "\n",
    "1. the [empirical distribution of $\\pi_W$](https://d-kwiat.github.io/gtg/within-host-pi.html) and how this can be used to estimate the quantum of transmission $Q$.\n",
    "\n",
    "1. the [relationship between $\\widehat{H}_W$ and $H_S$](https://d-kwiat.github.io/gtg/hw-hs.html).\n",
    "\n",
    "1. the [relationship between $F_{WS}$ and the parameters $\\chi$ and $Q$](https://d-kwiat.github.io/gtg/fws-compare-methods.html).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install coalestr\n",
    "from coalestr import cs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Observation time.    Events captured.   Mean coalescence time\n",
      "                      beho      wiho        beho     wiho\n",
      "        0              99.5     100.0     18188.4      1.0\n",
      "Observation time.  Nucleotide diversity     Haplotype homozygosity\n",
      "                      beho       wiho           beho       wiho\n",
      "        0           4.00e-04   2.20e-08       4.09e-03   9.87e-01\n"
     ]
    }
   ],
   "source": [
    "''' We create a history of transmission parameters ..\n",
    "\n",
    "my_history = [[D, N, Q, X, M]]\n",
    "    D = duration of simulation\n",
    "    N = effective number of hosts (Nh)\n",
    "    Q = quantum of transmission\n",
    "    X = crossing rate of transmission chains (chi)\n",
    "    M = number of migrant hosts (Nm) '''\n",
    "\n",
    "my_history = [[100000, 18764, 1, 0, 0]]\n",
    "my_population = cs.Population(my_history)\n",
    "my_population.get_coalescent()\n",
    "my_population.get_diversity()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the above example, $N_h = 18764$, $Q = 1$ and $\\chi = 0$.  The within-host parasite population has a mean coalescence time of 1 generation, giving nucleotide diversity $\\pi_W = 2.2 \\times 10^{-8}$ and haplotype homozygosity $\\gamma_W = 0.987$.\n",
    "\n",
    "In the example below $N_h = 18674$, $Q = 10$ and $\\chi = 1$.  The within-host parasite population has a mean coalescence time of 23517 generations, giving nucleotide diversity $\\pi_W = 5.17 \\times 10^{-4}$ and haplotype homozygosity $\\gamma_W = 0.0814$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Observation time.    Events captured.   Mean coalescence time\n",
      "                      beho      wiho        beho     wiho\n",
      "        0              60.5      63.8     25633.1  23517.1\n",
      "Observation time.  Nucleotide diversity     Haplotype homozygosity\n",
      "                      beho       wiho           beho       wiho\n",
      "        0           5.64e-04   5.17e-04       7.09e-04   8.14e-02\n"
     ]
    }
   ],
   "source": [
    "my_history = [[100000, 18764, 10, 1, 0]]\n",
    "my_population = cs.Population(my_history)\n",
    "my_population.get_coalescent()\n",
    "my_population.get_diversity()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
